Refactor `Typecore.type_application`

voodoos opened the pull request:
This PR proposes a refactor of `Typecore.type_application`.
This work was originally done by @lpw25 in Jane Street's fork of the compiler with extensions. Jane Street has asked for help from Tarides to upstream these changes.

This PR introduces a more descriptive type for arguments of a function application:

```ocaml
type expression_desc =
  | ...
  | Texp_apply of expression * (arg_label * apply_arg) list

and apply_arg = (expression, unit) arg_or_omitted

and ('a, 'b) arg_or_omitted =
  | Arg of 'a
  | Omitted of 'b
```

These were formerly represented with the type `expression option`.

The typing of applications is now clearly split in two phases: first `collect_apply_args` maps over the arguments, classify them and collect additional information. Then `type_apply_arg` and `type_omitted_parameters` perform the actual typing.

The classification was previously done by imperatively filling lists such as `eliminated_optional_arguments` and `omitted_parameters`. Now it uses the type parameters of `arg_or_omitted` to store additional information and specific type constructors are introduced for these cases:

```ocaml
type untyped_apply_arg =
  | Known_arg of
      { sarg : Parsetree.expression;
        ty_arg : type_expr;
        ty_arg0 : type_expr;
        mode_arg : Alloc_mode.t;
        wrapped_in_some : bool; }
  | Unknown_arg of
      { sarg : Parsetree.expression;
        ty_arg : type_expr;
        mode_arg : Alloc_mode.t; }
  | Eliminated_optional_arg of
      { mode_fun: Alloc_mode.t;
        ty_arg : type_expr;
        mode_arg : Alloc_mode.t;
        level: int;}

type untyped_omitted_param =
  { mode_fun: Alloc_mode.t;
    ty_arg : type_expr;
    mode_arg : Alloc_mode.t;
    level: int; }

(untyped_apply_arg, untyped_omitted_param) arg_or_omitted
```

Thus, the flow for typing an application can now be summarized with these three functions:

```ocaml
collect_apply_args :
  ... (arg_label * expression) list ->
  ... * (untyped_apply_arg, untyped_omitted_param) arg_or_omitted list

type_apply_arg :
  ... (untyped_apply_arg, untyped_omitted_param) arg_or_omitted ->
  ... * (Typedtree.expression, untyped_omitted_param) arg_or_omitted

type_omitted_parameters :
  ... (Typedtree.expression, untyped_omitted_param) arg_or_omitted list ->
  ... * (Typedtree.expression, unit) arg_or_omitted list
```

I was not familiar with the typing process of function application before doing this work but my impression is that these changes do make the code clearer and more robust.

cc @OlivierNicole @goldfirere

lpw25 committed:
Refactor `Typecore.type_application`

The typing of applications is now decomposed in two phases: collecting the
arguments and typing them.

Co-authored-by: Ulysse Gérard <thevoodoos@gmail.com>

MisterDA commented on ocamldoc/odoc_ast.ml:
```diff
@@ -725,12 +725,12 @@ module Analyser =
-                match exp_opt with
-                  None -> acc
-                | Some e -> acc @ [e])
+              (fun acc -> fun (_, arg) ->
+                match arg with
+                | Omitted () -> acc
+                | Arg e -> acc @ [e])
               []
-              exp_opt_optional_list
+              arg_list
           in
```
isn't that just `List.filter (fun (_, arg) -> not (is_omitted arg)) arg_list` (provided you carry the definition of `is_omitted` here?

voodoos replied:
Actually, `filter_map` is the fitting one here. I refactored it in [183948c](https://github.com/ocaml/ocaml/pull/13612/commits/183948c289df28e5150af502fb84f1fc252a2709)

goldfirere commented:
Thanks, @voodoos, for posting this!

voodoos committed:
Refactor: use std function to filter a list in odoc_ast

Suggested-by: Antonin Décimo <antonin@tarides.com>

voodoos committed:
Add changelog entry for #13612

gasche commented on lambda/translcore.ml:
```diff
@@ -213,10 +217,10 @@ and transl_exp0 ~in_new_scope ~scopes e =
       let argl, extra_args = cut p.prim_arity oargs in
       let arg_exps =
-         List.map (function _, Some x -> x | _ -> assert false) argl
+         List.map (function _, Arg x -> x | _ -> assert false) argl
```
I would write `| Omitted () -> assert false` which is more precise/robust -- it clarifies why the `not (is_omitted arg)` test above is enough to prevent this case.

gasche commented on lambda/translcore.ml:
```diff
@@ -714,7 +718,12 @@ and transl_apply ~scopes
         (* Evaluate the remaining arguments;
            if we already passed here this is a no-op. *)
         let l =
-          List.map (fun (arg, opt) -> Option.map (protect "arg") arg, opt) l
```
I wonder if it would make sense to introduce `Typedtree.map_apply_arg` to replace `Option.map`, its absence creates some less-necessary code churn in this module (4 occurrences of `Option.map` have been inlined manually).

gasche reviewed (commented):
(A very shallow first review on the small part of the change that is not within `typecore.ml`, which the github UI hides by default.)

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+      {
+        ty_arg : type_expr;
+        level: int;
+      }
```
As someone who is unfamiliar with the previous state of the codebase, it is not obvious what these things mean. Could you consider writing a few documentation comments that give concrete examples of `Known_arg`, `Unknown_arg`, and `Eliminated_optional_arg` arguments? Other questions I would naively ask myself are: what is the difference between `ty_arg` and `ty_arg0`, what `wrapped_in_some` means (maybe examples with and without), and what the `level` is.

samsa1 replied:
`wrapped_in_some` corresponds to an optional argument application :
```ocaml
let f : ?x:_ -> _
let _ = f ~x
```
When typing the application, `x` will be stored in a `Known_arg` with `wrapped_in_some = true` if I understand correctly (I definitely agree that comments should be added at least for future contributors)

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
     -> final_subexpression e
   | _ -> exp
 
+let is_prim ~name funct =
```
For reviewers: this function and `list_labels{,_aux}` below are just moved upwards from where they were in the file.

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+    level: int;
+  }
+
+let remaining_function_type ty_ret rev_args =
```
For reviewers: this corresponds to `result_type` at the beginning of `type_application` in the trunk implementation, except that it takes the full (rev)list of `apply_arg`, not just omitted and eliminated arguments.

garrigue replied:
I find this function particularly confusing, as it mixes two steps.
The original `result_type` was just rebuilding a type, given labels and levels.
It was called in two distinct places:
* in the default case of `type_unknown_arg`, in order to generate an error message. In that case we need to pretend that omitted arguments are still there, since the error may be that we misspelled a label so that it ended up being omitted.
* in `type_unknown_args`, in order to generate the final type of the function application. In that case omitted arguments are of course omitted.

But this version is actually specialized for the first occurrence, which explains why it handles omitted arguments as kept.
At least it should be renamed to `remaining_function_type_for_error`.

Looking at code, I also wonder if your choice of types is ideal.
Would it not be better to reserve `Arg` for actual arguments, and put erased arguments on the `Omitted` side. 
```ocaml
type untyped_omitted_param =
  {
    eliminated : bool ;  (* true for eliminated optional arguments, false otherwise *)
    ty_arg : type_expr;
    level: int;
  }
```
Since this type is internal to `type_application`, there is no external impact, and it could clarify the code.

lpw25 replied:
> Would it not be better to reserve Arg for actual arguments, and put erased arguments on the Omitted side.

It is much more important to distinguish omitted parameters -- which cause a closure to be created rather than a function to be called -- from ordinary arguments. Whist `untyped_omitted_param` is local to `type_application` the use of the `arg_or_omitted` is not: the distinction between actual arguments and omitted parameters is preserved all the way through to `translcore`. Getting this distinction clear and correct was a key motivation for this patch (it matters *a lot* for our work on modes like `local`).

garrigue replied:
Indeed, `arg_or_omitted` is kept. But `untyped_apply_arg` is not.
And I am still under the impression that, until they get typed, the code would be simpler with eliminated optional arguments on the `Omitted` side. In particular, in this function the distinction between omitted arguments that are eliminated and those that are kept does not matter yet. Of course, you need to move them to the `Arg` side eventually, when you create the concrete argument. This can be done either in `type_apply_arg` or `type_omitted_parameters`. I would actually prefer the latter, and it would even give you a justification for the name.

garrigue replied:
Sorry, actually this is not what I wanted to say.
Indeed, `arg_or_omitted` is kept. But `untyped_apply_arg` is not.
And I am still under the impression that, until they get typed, the code would be simpler with eliminated optional arguments on the `Omitted` side. In particular in this function, the point being that there are two kinds of omitted arguments, those that get eliminated and those that are kept, but it only matters later on. Of course, you need to move them to the `Arg` side when you create the concrete argument. Either in `type_apply_arg` or `type_omitted_parameters`; I would actually prefer the later, and it would even give you a justification for the name.

lpw25 replied:
Having the meaning of `Omitted` be inconsistent between different parts of the code just makes it much harder to understand for no obvious benefit. Eliminated optional parameters are fundamentally *not* omitted arguments and it is confusing to label them as such.

garrigue replied:
But they are omitted. Seeing them otherwise makes this function hard to understand.
Anyway, I don't want to make this a requirement (this is why I approved).
This function still must be renamed: if you see eliminated optional parameters as not omitted, they clearly cannot be part of the remaining function type.

My point of view is that `type_application` is converting some omitted arguments into defaults, the point at which it is done in the function being a question of design choice. The benefit of doing it at the end is that the intermediate representation becomes closer to the source syntax, which for instance makes reasoning about errors easier.

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+            ty_ret)
+    ty_ret rev_args
+
+let collect_unknown_apply_args env funct ty_fun rev_args sargs =
```
For reviewers: this corresponds more or less to the `type_unknown_arg` helper of the trunk implementation. `labels_match` below is a new auxiliary function, and `has_label` is moved here from the beginning of `type_application`.

Note that `type_unknown_arg` types a single `(lbl, sarg)` argument and adds it to its `typed_args` argument, while the present function processes a list of arguments at once, closer to `type_unknown_args ()` in the original implementation.

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+  in
+  loop ty_fun rev_args sargs
+
+let collect_apply_args env funct ignore_labels ty_fun ty_fun0 sargs =
```
For reviewers: this corresponds to the `type_args` function within the trunk `type_application` function.

gasche commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+  in
+  loop ty_fun ty_fun0 [] sargs
+
+let type_omitted_parameters ty_ret args =
```
Still trying to map the new code to the old code: it is unclear to me where this function comes from, it appears to be new. There seems to be some redudancy with the function `remaining_function_type` above, I am not sure why this is.

voodoos replied:
Part of the code for this function actually comes from the `result_type` auxiliary function defined at the beginning of trunk's `type_application` function and which is called over the list of `!omitted_parameters`.

gasche replied:
But `result_type` was already moved into `remaining_function_type` above. Why have the same Tarrow-building logic in two different places?

samsa1 replied:
For me the difference is that :
- `remaining_function_type` of `(f : ?x:int -> y:bool -> float -> t) 3.` is `?x:int -> y:bool -> t` because `?x` will go in the `Arg (Eliminated_optional_arg _)` case.
- `type_omitted_parameters` of `(f : ?x:int -> y:bool -> float -> t) 3.` is `y:bool -> t`
The first case is used for error messages when the second is the resulting type after eta-expanding ommited labelled arguments.
It might be nice to put both function together with a comment to explain why they are different.

Octachron replied:
We could merge the two function by adding an `?reinject_eliminated_optional_arg` arguments to `type_omitted_parameters` and discarding the argument part in the error path that was using `remaining_function_type`.

I also think that it could be nice the rename the function to `return_function_type` or `return_type_with_omitted_parameters` to better convey the idea that there is no typing going on in the function `type_omitted_parameters`.

garrigue replied:
Indeed, this is the second usage of `result_type` in the original code (see my comment above).
It also builds the list of desugared arguments, so clearly the name is wrong.
Maybe `build_result_type_and_arguments`.

lpw25 replied:
The reason it is called `type_omitted_parameters` is that it maps over the omitted parameters turning them from their untyped form -- `untyped_omitted_param` -- into their typed form. Unfortunately, in this patch the typed form is just `unit` so it looks a bit weird. Our work on both modes and kinds requires us to put more information into the typed form, so that we have a type `omitted_param` in `Typedtree` and that is what is produced for the omitted parameters here.

I do think this is still to right way to think about this function however, so I'd prefer if this view could be preserved in the name.

It is true than this function is really a `fold_map` rather than a `map` -- it also computes the `result_type` as it goes -- and this isn't reflected in the current name. It probably would be better if the name included that detail. How about something like `type_omitted_parameters_and_build_result_type`?

garrigue replied:
Here I really think that this PR should use a name that corresponds to what is done, not what will be done in the future. You can easily change this name when the meaning changes: this function is used only once.

gasche replied:
My understanding is that it's not just about a change in the future, it about reducing the diff between upstream and Jane Street's _existing_ fork of the type-checker codebase. I think that it's worth accommodating this need if we can at reasonable cost, to make such upstreaming easier (and worthy) in the future. Maybe we could find a name that would be acceptable in both versions, for example `handle_omittted_parameters_and_build_result_type`?

lpw25 replied:
My point was actually slightly stronger than just about keeping the diff small. I was also arguing that semantically this is the place where we "type" the omitted arguments: taking them from their representation in the parsetree to their representation in the typedtree. Just because the representation in the typedtree is `()` doesn't mean it is the wrong way to picture this function.

Still, I would settle for `handle_omitted_parameters_...` if people object to using the word `type_` for this.

garrigue replied:
If you want to call it that way, that's fine.
For a preciser picture, we would need to see the real code.

gasche reviewed (commented):
I tried to have a look at the diff as a neophyte, and started by trying to piece out a map from the new code (which comes first in the github diff) to the old code. If I wanted to review this, I guess that I would try to check that the newly-located code is indeed equivalent to the previous code, and separately try to read the new body of the `type_application` function and compare with the previous one.

gasche commented:
(Note: the fact that the change comes in one big commit makes it harder to review than if it was split in a sequence of smaller changes.)

samsa1 commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+    | Tarrow (l, ty_arg, ty_ret, com), Tarrow (_, ty_arg0, ty_ret0, _)
+      when is_commu_ok com ->
+        let lv = get_level ty_fun' in
+        let may_warn loc w =
+          if not !warned && !Clflags.principal && lv <> generic_level
+          then begin
+            warned := true;
+            Location.prerr_warning loc w
+          end
+        in
```
If we add `lv` as a parameter of this function we can move it outside of the `loop` function. Thus closed to the definition of `warned` that could even become local to this function definition.

samsa1 replied:
But it might be better to minimize the diff and leave such rewriting for another time

garrigue replied:
I am not sure it would be correct. This function iterates on the spine of the function type, and the level changes along it.

samsa1 commented on typing/typecore.ml:
```diff
@@ -2596,6 +2596,238 @@ let rec final_subexpression exp =
+        in
+        loop ty_ret ty_ret0 ((l, arg) :: rev_args) remaining_sargs
+    | _ ->
+      (* We're not looking at a *known* function type anymore, or there are no
+        arguments left. *)
```
This comment comes from the previous definition of `type_unknown_args` and is partially out of date because we know that we have at least one argument left (see `if` on line 2732)

goldfirere commented:
I offered to help out with review here. What's the current status of review? How can I be of service?

gasche commented on typing/typedtree.mli:
```diff
@@ -210,7 +210,7 @@ and expression_desc =
             Texp_apply (Texp_ident "f/1037",
-                        [(Nolabel, None);
+                        [(Nolabel, Omitted ());
                          (Labelled "y", Some (Texp_constant Const_int 3))
```
Some => Arg

voodoos committed:
type_args: move [expand_head] to match trunk evaluation order

gasche committed:
typecore.ml: documentation comments for type_application

Co-authored-by: Samuel Vivien <samuel.vivien@inria.fr>
Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>
Co-authored-by: Ulysse Gérard <thevoodoos@gmail.com>

gasche committed:
typecore.ml: small location improvement in an type_application error message

Co-authored-by: Samuel Vivien <samuel.vivien@inria.fr>
Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>
Co-authored-by: Ulysse Gérard <thevoodoos@gmail.com>

gasche committed:
typecore.ml: documentation comments for untyped_apply_arg

Co-authored-by: Samuel Vivien <samuel.vivien@inria.fr>
Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>
Co-authored-by: Ulysse Gérard <thevoodoos@gmail.com>

gasche committed:
typecore.ml: an [instance] was possibly missing compared to the trunk version

(This might make a difference when reconstructed omitted parameters that were deconstructed from a generic type.)

Co-authored-by: Samuel Vivien <samuel.vivien@inria.fr>
Co-authored-by: Florian Angeletti <florian.angeletti@inria.fr>
Co-authored-by: Ulysse Gérard <thevoodoos@gmail.com>

samsa1 committed:
fix a comment

gasche committed:
minor renaming for consistency

gasche commented:
@voodoos, @sama1, @Octachron and myself reviewed this commit together. We pushed a string of changes as we went along. We are now collectively convinced that the new behavior is the same as the previous one.

(@samsa1 said he would look at the location we use in error messages for eliminated optional arguments)

gasche commented on Changes:
```diff
@@ -139,6 +139,9 @@ Working version
   (Mark Shinwell, review by Vincent Laviron)
 
+- #13612: Refactor `type_application`
+  (Ulysse Gérard, Leo White, review by Antonin Décimo, ...)
```
(This needs an update.)

gasche reviewed (approved):
Approved in principle.

@samsa1 is wondering about error message locations, and @voodoos should do a pass on the review comments and see if minor changes need to be applied. (I need to rebase to remove the `fixup!` commit.)

samsa1 commented:
I managed to convince myself that there was no problem with location.
I approve of the changes on this PR.


For the record :

https://github.com/ocaml/ocaml/blob/061adb721511a5d387e67c56abbfa57c34e9efe0/typing/typecore.ml#L5479

and

https://github.com/ocaml/ocaml/blob/061adb721511a5d387e67c56abbfa57c34e9efe0/typing/typecore.ml#L5499

Where incoherent between classic and non-classic mode. This could lead to an difference in error messages. This PR chooses the approach of not storing the location (which is understandable because there are no corresponding arguments). However this could have impacted locations in error messages.

However I found only one usage of the location stored in that list :
https://github.com/ocaml/ocaml/blob/061adb721511a5d387e67c56abbfa57c34e9efe0/typing/typecore.ml#L5355-L5382

However because of the match on the `result_type` we can never go into the second case of the `match` if we had ommited an optionnal argument before.
Thus the question of which location is stored for the eliminated optional argument is irrelevant because it is never used.

gasche committed:
typecore.ml: minor tweak

gasche commented:
( @goldfirere sorry for the lack of responsiveness here; last week was my "avoid looking at github/ocaml notifications" week and I missed your question. I think that we can move to merge this quickly, and maybe spend the coordination effort on the review of polymorphic arguments. )

garrigue reviewed (changes requested):
I did not check all the details, but the approach seems fine.
Being more functional is always good, but the main goal seems here to be the separation between argument selection and typing.
I am a bit curious of how all that interacts with Merlin (I suppose that you already checked).
My comments are mostly stylistic, but I think it matters here.
Particularly, the two functions that replace `result_type` should be at least renamed and their roles explained.

garrigue reviewed (approved):
I am happy with the code, just need to clarify function names.

voodoos committed:
Introduce `Typedtree.map_apply_arg`

Suggested-by: Gabriel Scherer <gabriel.scherer@gmail.com>

voodoos committed:
Reword some comments.

voodoos committed:
Rename `type_omitted_parameters`

voodoos committed:
Update reviewers list in changelog

voodoos commented:
I made the requested changes. I chose to rename `type_omitted_parameters` to `type_omitted_parameters_and_build_result_type` (following the argument saying that "this is the place where we "type" the omitted arguments: taking them from their representation in the parsetree to their representation in the typedtree")

I can switch it promptly if you prefer  `handle_omitted_parameters_and_build_result_type` !

gasche commented on lambda/translcore.ml:
```diff
@@ -719,10 +719,7 @@ and transl_apply ~scopes
-               match arg with
-               | Omitted () -> arg, opt
-               | Arg arg -> Arg (protect "arg" arg), opt)
+            (fun (arg, opt) -> Typedtree.map_apply_arg (protect "arg") arg, opt)
```
(I think this change and the one below were meant to go in the *previous* commit, but it's no big deal and fine to leave as-is.)

voodoos replied:
I decided to defy Flemme and made the change :-)

gasche commented:
I am happy with the resulting PR, and would be in favor of merging as-is because it's good enough.

(The history is slightly messy now, with several small review commits, but I also propose to leave it as-is because flemme.)

voodoos committed:
Make matching exhaustive

gasche commented:
Merged! Thanks everyone for your time.

garrigue commented:
It seems that this was merged without renaming `remaining_function_type`, which is clearly wrong.
Will have to fix it.

voodoos commented:
I must have missed it when re-reading the comments. I just made the change you suggested in #13715
